### DISCUSSION TITLE
로그 프로토콜에 따른 데이터값 해석방법 및 GPS지도 안뜨는 현상

### DISCUSSION THREAD

**Discussion by Stella-repo** [luftaquila/monolith#18] [Q&A] [answered]
안녕하세요 창원대학교 SKID 입니다.

현재 모노리스를 통해 1차적으로 차량의 정보를 받고 UART나 I2C통신을 통해 로그데이터를 받아서 대시보드의 디스플레이를 출력하려고 합니다.
그래서 로그데이터를 우리가 알아볼 수 있는 데이터로 바꾸고자 하는데 어려움이 있어서 질문드립니다.

다음은 I2C통신을 통해 아두이노로 모노리스의 로그데이터를 받은 값과 SD카드에 저장된 로그데이터를 TMA-3의 로그변환기를 통해 변환한 값들입니다.

값 1: 0x2096
값 2: 0x67000503
값 3: 0xFFFF0009
값 4: 0x00A2
  {
    "timestamp": 8442,
    "datetime": "2024-03-05 13:43:29.442",
    "level": "INFO",
    "source": "ACCELEROMETER",
    "key": "ACCELEROMETER_DATA",
    "checksum": true,
    "value": 704374571017,
    "raw": {
      "0": 9,
      "1": 0,
      "2": 255,
      "3": 255,
      "4": 163,
      "5": 0,
      "6": 0,
      "7": 0
    },
    "parsed": {
      "x": 0.0703125,
      "y": -0.0078125,
      "z": 1.2734375
    }
  },

값 1: 0x20F4
값 2: 0x11010303
값 3: 0x069703A7
값 4: 0x09EF09AE
  {
    "timestamp": 8436,
    "datetime": "2024-03-05 13:43:29.436",
    "level": "INFO",
    "source": "ANALOG",
    "key": "ANALOG_DATA",
    "checksum": true,
    "value": 715801508814717800,
    "raw": {
      "0": 167,
      "1": 3,
      "2": 151,
      "3": 6,
      "4": 174,
      "5": 9,
      "6": 239,
      "7": 9
    },
    "parsed": {
      "AIN0": 0.22832722832722832,
      "AIN1": 0.41196581196581195,
      "AIN2": 0.6051282051282051,
      "AIN3": 0.621001221001221
    }
  },

여기서 가속도 데이터는 2의 보수를 적용하여 10진수로 변환한 후 *4/512를 적용하면 된다는것은 이해했습니다.
그런데 아날로그 데이터는 어떻게 변환을 적용해야할지 잘 모르겠습니다.

그리고 추가적으로 실시간 텔레메트리 화면에서 GPS지도가 안뜨는 현상이 있었습니다.
첨부한 사진에는 현재 서버에 연결되어있지 않은것으로 뜨는데 서버가 연결된 상황에서도 똑같은 화면이 떴었습니다.
그리고 폰, 태블릿, PC여러대에서 다 확인했을때도 동일한 증상이 발생했었습니다.
추가적으로 SD카드에 저장된 로그데이터를 불러와서 볼때는 정상적으로 GPS지도가 뜨는것을 확인했습니다.

![Cap 2024-03-06 16-22-10-050](https://github.com/luftaquila/monolith/assets/93361308/0d967a6a-7760-494d-b7d8-8152a3b5010e)
![Cap 2024-03-06 16-22-43-139](https://github.com/luftaquila/monolith/assets/93361308/95de64d2-3e19-49c1-856f-ca8428942ca8)

감사합니다.

**Comment by luftaquila** [comment]
우선 GPS 지도가 안 뜬다는 말씀은 GPS 좌표가 안 찍히는게 아니라 지도 자체가 표시가 제대로 안 된다는 말씀이신 거죠?

**Reply by Stella-repo** [reply]
네 맞습니다
GPS좌표는 실시간으로 뜨는것 확인했습니다

**Reply by luftaquila** [reply]
지도가 안 뜨는 이유는 확인이 안 되네요. 브라우저 / OS 정보나 F12 누르면 나오는 개발자 도구 콘솔에 에러 찍힌게 있는지 정도를 알려주시면 도움이 될 것 같아요. 제 브라우저 몇 개에서는 정상적으로 동작해서요.

혹시 처음에는 잘 나오다가 GPS 데이터가 들어오기 시작하면 깨진다거나 그런 현상이 있나요? 조금 더 상세하게 알려 주시면 도움이 될 듯합니다.

**Reply by luftaquila** [reply]
안녕하세요,

지도 안 뜨는 현상 원인을 발견해 수정했습니다. (#22)

GPS 데이터 미수신 중에는 문제가 없으나, 데이터가 들어오는 중에 좌표값이 잘못 전달되어 GPS 지도가 깨지는 문제가 있었습니다.

https://monolith.luftaquila.io/live/ 에도 수정된 버전이 반영되었습니다. 제보 감사합니다.

**Reply by Stella-repo** [reply]
최근에 테스트를 못해봤었는데 수정해주셔서 감사합니다.
테스트 해보았을때 문제 없이 GPS지도가 뜨는것 확인했습니다.

**Comment by luftaquila** [comment, marked as answer]
데이터 해석 답변이 늦어서 죄송합니다. 혹시 [로그 프로토콜](https://github.com/luftaquila/monolith/wiki/%EB%A1%9C%EA%B7%B8-%ED%94%84%EB%A1%9C%ED%86%A0%EC%BD%9C) 문서를 확인해 보셨나요?

데이터 자체는 잘 수신이 되고 있는 것으로 보입니다. 다만 첫 번째 로그는 `값 1`이 0x00002096 이라면 timestamp가 8442가 아니라 8342가 되어야 할 것 같네요.

![image](https://github.com/luftaquila/monolith/assets/17094868/df7ab305-892c-404f-acee-75cbf4f185e7)

아래 이미지에서 각 행이 `값 #` 와 일치한다고 생각하시면 됩니다. `값 1` 은 timestamp이고, `값 2`는 level, source, key, checksum을 포함하고 있으며, `값 3`과 `값 4`가 value 8바이트를 나타냅니다.

다만 해석할 때 endianness를 고려해야 하는데요, 두 번째 예시의 `값 2` 0x11010303 를 예로 들어 보겠습니다.

```
0x  11  01  03  03
    ^^  **  --  ##
```

가장 낮은 8비트 (##) 은 level로, 0x03은 로그 프로토콜 문서를 참조하시면 `INFO` 가 3이라는 것을 알 수 있습니다.

두 번째 8비트 (--)인 source는 마찬가지로 0x03이므로 `ANALOG` 에 해당합니다.

세 번째 8비트 (**)인 key는 0x01인 것을 보면 `ANALOG_DATA` 라는 것을 알 수 있습니다.

마지막 checksum (^^) 인 0x11은 정해진 규칙 `(자신을 제외한 나머지 15개 바이트의 합) AND (0xFF)` 에 따라 계산된 값이고요.

<hr>

[ANALOG_DATA](https://github.com/luftaquila/monolith/wiki/%EB%A1%9C%EA%B7%B8-%ED%94%84%EB%A1%9C%ED%86%A0%EC%BD%9C#analog_data) 해석은 문서를 보시면 value의 0, 1번째 byte가 `AIN0` 을 의미한다고 되어 있습니다. 올려주신 두 번째 로그의 `값 3` 0x069703A7 에서 0x03A7 입니다.

이 값은 10진수로 935이고, 이는 0~4095 범위의 8비트 ADC 변환 결과값입니다. 따라서 `935 / 4095 = 0.228327` 이 로그에 기록됩니다. 이 값에 STM32 ADC의 작동 전압인 3.3V를 곱한 `0.75347V` 가 STM32에 입력되고 있는 전압입니다.

이걸로 실제 `AIN0` 단자에 입력되는 전압을 계산하려면 [TMA-1 DIY 가이드 ANALOG 영역](https://github.com/luftaquila/monolith/wiki/TMA%E2%80%901-DIY-%EA%B0%80%EC%9D%B4%EB%93%9C#2-5-analog-%EC%98%81%EC%97%AD) 에서 PCB에 설정하신 R25-R32 저항 값을 사용하면 됩니다.

만약 가이드대로 PCB ANALOG 영역의 홀수 번호 저항에 470옴, 짝수 번호저항에 910옴을 다셨다면, `0.75347 * (470 + 910) / 910 = 1.142639V` 가 실제 `AIN0` 단자의 전압이 됩니다.

설명하느라 길어졌는데요, 요약하면

`(해당 단자 2바이트 값) / 4095 * 3.3 * (홀수 번호 저항값 + 짝수 번호 저항값) / (짝수 번호 저항값)` 이 V 단위의 실제 입력 전압이 됩니다.

올려 주신 로그의 value에 해당하는 

```
값 3: 0x069703A7
값 4: 0x09EF09AE
```

에서 `AIN0` 단자 전압은 위 식의 `해당 단자 2바이트 값`에 0x03A7을, `AIN1`은 0x0697, `AIN2` 는 0x09AE, `AIN3` 은 0x09EF 를 넣으면 됩니다.

parsed 값으로 나오는 건 `(해당 단자 2바이트 값) / 4095` 까지만 계산한 값이고요.

**Reply by luftaquila** [reply]
덧붙이면 `AIN0~3` 단자에 아무것도 연결되지 않은 floating 상태라면 원래 값이 랜덤하게 튑니다. 아날로그 단자가 floating인 건 undefined behavior 라서요. TMA-1과 그라운드를 공유하는 전압원을 연결해 주시면 제대로 출력됩니다.

**Reply by Stella-repo** [reply]
친절하고 상세하게 답변주셔서 감사합니다. 덕분에 완전히 이해가 된것같습니다.
ANALOG_DATA 해석에 있어서 parsed값이 전압값이라고 생각을 해서 계속 헤맸던것 같습니다.

GPS지도에 대한 부분은 제가 조금 더 다양하게 확인해보고 추가로 알려드리겠습니다.
감사합니다.
